internal/transports/p2p/peer/peer.go

```diff
@@ -0,0 +1,544 @@
+package peer
+
+import (
+	"crypto/rand"
+	"errors"
+	"fmt"
+	"math"
+	"math/big"
+	"net"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/bitcoin-sv/block-headers-service/config"
+	"github.com/bitcoin-sv/block-headers-service/domains"
+	"github.com/bitcoin-sv/block-headers-service/internal/chaincfg"
+	"github.com/bitcoin-sv/block-headers-service/internal/chaincfg/chainhash"
+	"github.com/bitcoin-sv/block-headers-service/internal/wire"
+	"github.com/bitcoin-sv/block-headers-service/service"
+	"github.com/rs/zerolog"
+)
+
+type Peer struct {
+	conn            net.Conn
+	addr            *net.TCPAddr
+	cfg             *config.P2PConfig
+	chainParams     *chaincfg.Params
+	checkpoints     []chaincfg.Checkpoint
+	nextCheckpoint  *chaincfg.Checkpoint
+	headersService  service.Headers
+	chainService    service.Chains
+	log             *zerolog.Logger
+	services        wire.ServiceFlag
+	protocolVersion uint32
+	nonce           uint64
+	lastBlock       int32
+	timeOffset      int64
+	userAgent       string
+	synced          bool
+
+	wg       sync.WaitGroup
+	msgChan  chan wire.Message
+	quitting bool
+	quit     chan struct{}
+}
+
+func NewPeer(
+	addr string,
+	cfg *config.P2PConfig,
+	chainParams *chaincfg.Params,
+	headersService service.Headers,
+	chainService service.Chains,
+	log *zerolog.Logger,
+) (*Peer, error) {
+	port, err := strconv.Atoi(config.ActiveNetParams.DefaultPort)
+	if err != nil {
+		return nil, fmt.Errorf("could not parse default port: %w", err)
+	}
+
+	ip := net.ParseIP(addr)
+	if ip == nil {
+		return nil, errors.New("could not parse peer IP")
+	}
+
+	netAddr := &net.TCPAddr{
+		IP:   ip,
+		Port: port,
+	}
+
+	currentTipHeight := headersService.GetTipHeight()
+	nextCheckpoint := findNextHeaderCheckpoint(chainParams.Checkpoints, currentTipHeight)
+
+	peer := &Peer{
+		addr:            netAddr,
+		cfg:             cfg,
+		chainParams:     chainParams,
+		headersService:  headersService,
+		chainService:    chainService,
+		checkpoints:     chainParams.Checkpoints,
+		nextCheckpoint:  nextCheckpoint,
+		log:             log,
+		services:        wire.SFspv,
+		protocolVersion: initialProtocolVersion,
+		synced:          nextCheckpoint == nil,
+		wg:              sync.WaitGroup{},
+		msgChan:         make(chan wire.Message, writeMsgChannelBufferSize),
+		quitting:        false,
+		quit:            make(chan struct{}),
+	}
+	return peer, nil
+}
+
+func (p *Peer) Connect() error {
+	conn, err := net.Dial(p.addr.Network(), p.addr.String())
+	if err != nil {
+		p.log.Error().Msgf("error connecting to the peer, reason: %v", err)
```

Comment:
It would be great to have peer info here also:
```suggestion
		p.log.Error().Msgf("error connecting to the peer %s, reason: %v", p, err)
```

Reply:
Right, I missed it, done